usr.bin/mdo/mdo.c

```diff
@@ -19,52 +20,277 @@
+			if (strspn(username, "0123456789") == strlen(username)) {
+				const char *errp = NULL;
+				uid_t uid = strtonum(username, 0, UID_MAX, &errp);
+				if (errp != NULL)
+					err(EXIT_FAILURE, "invalid user ID '%s'", username);
+				pw = getpwuid(uid);
+			}
+			if (pw == NULL)
+				err(EXIT_FAILURE, "invalid username '%s'", username);
 		}
-		if (pw == NULL)
-			err(EXIT_FAILURE, "invalid username '%s'", username);
 	}
 
 	wcred.sc_uid = wcred.sc_ruid = wcred.sc_svuid = pw->pw_uid;
 	setcred_flags |= SETCREDF_UID | SETCREDF_RUID | SETCREDF_SVUID;
 
-	if (!uidonly) {
+	if (ruid_str) {
+		const char *errp = NULL;
+		wcred.sc_ruid = strtonum(ruid_str, 0, UID_MAX, &errp);
+		if (errp)
+			err(EXIT_FAILURE, "-U: invalid UID");
+		setcred_flags |= SETCREDF_RUID;
+	}
+	if (svuid_str) {
+		const char *errp = NULL;
+		wcred.sc_svuid = strtonum(svuid_str, 0, UID_MAX, &errp);
+		if (errp)
+			err(EXIT_FAILURE, "-R: invalid UID");
+		setcred_flags |= SETCREDF_SVUID;
+	}
+	if (euid_str) {
+		const char *errp = NULL;
+		wcred.sc_uid = strtonum(euid_str, 0, UID_MAX, &errp);
+		if (errp)
+			err(EXIT_FAILURE, "-E: invalid UID");
+		setcred_flags |= SETCREDF_UID;
+	}
+	if (rgid_str) {
+		const char *errp = NULL;
+		wcred.sc_rgid = strtonum(rgid_str, 0, GID_MAX, &errp);
+		if (errp)
+			err(EXIT_FAILURE, "-P: invalid GID");
+		setcred_flags |= SETCREDF_RGID;
+	}
+	if (svgid_str) {
+		const char *errp = NULL;
+		wcred.sc_svuid = strtonum(svgid_str, 0, GID_MAX, &errp);
+		if (errp)
+			err(EXIT_FAILURE, "-Q: invalid GID");
+		setcred_flags |= SETCREDF_SVGID;
+	}
+
+	if (primary_group) {
+		struct group *gr = getgrnam(primary_group);
+		if (gr)
+			gid = gr->gr_gid;
+		else {
+			const char *errp = NULL;
+			gid = strtonum(primary_group, 0, GID_MAX, &errp);
+			if (errp != NULL)
+				err(EXIT_FAILURE, "invalid group '%s'", primary_group);
+		}
+		override_gid = true;
+	} else if (!uidonly && pw) {
+		gid = pw->pw_gid;
+		override_gid = true;
+	}
+
+	if (override_gid) {
+		wcred.sc_gid = wcred.sc_rgid = wcred.sc_svgid = gid;
+		setcred_flags |= SETCREDF_GID | SETCREDF_RGID | SETCREDF_SVGID;
+	}
+
+	if (supp_groups_str) {
+		char *copy = strdup(supp_groups_str);
+		if (!copy)
+			err(EXIT_FAILURE, "strdup failed");
+		char *tok = strtok(copy, ",");
+		while (tok) {
+			struct group *gr = getgrnam(tok);
+			gid_t g;
+			if (gr)
+				g = gr->gr_gid;
+			else {
+				const char *errp = NULL;
+				g = strtonum(tok, 0, GID_MAX, &errp);
+				if (errp != NULL)
+					err(EXIT_FAILURE, "invalid group '%s", tok);
+			}
+			supp_add = realloc(supp_add, sizeof(gid_t) * (add_count + 1));
+			if (!supp_add)
+				err(EXIT_FAILURE, "realloc failed");
+			supp_add[add_count++] = g;
+			tok = strtok(NULL, ",");
+		}
+		free(copy);
+		supp_reset = true;
+	}
+
+	if (group_mod_str) {
+		char *s = strdup(group_mod_str);
+		if (!s)
+			err(EXIT_FAILURE, "strdup failed");
+		char *tok = strtok(s, ",");
+		while (tok) {
+			if (strcmp(tok, "@") == 0) {
```

Comment:
I think it would be better if `@` is accepted only as the first token.  Else, there is confusion between whether `@` should reset everything first even when not in first position, or should reset groups after the first directives, which anyway would be silly (and I don't see a use case for this feature with automatically produced command lines).